### ISSUE TITLE
Need a more flexible way to specify benchmark options

### ISSUE THREAD

**Issue by atheurer** [distributed-system-analysis/pbench#531]
When we run a benchmark, all of the options used apply to every test type.  For example, --protocols=tcp,udp will be used for all --test-types=stream,maerts,bidirec,rr.  If the user wanted only udp for test type rr, but only tcp for test type stream, they would need to run pbench-uperf twice.

We need a way to describe all of the tests we want, but have different options for different set of tests.  Yaml may be the easiest way to do this.

**Comment by atheurer** [comment]
Here is a YAML example that might work for uperf:

'
- one set of tests I want:
  - test-types:
    - stream
    - mearts
    - bidirec
  - message-sizes:
    - 64
    - 1024
    - 16384
  - instances:
    - 1
    - 16
- another set of tests I want:
  - test-type: [rr]
  - message-sizes: [[1,256],[1,16384]]
  - instances: [64]
'

**Comment by jtaleric** [comment]
I like the highly nested YAML. I find it the most flexible, however maybe not the most readable.

**Comment by portante** [comment]
YAML seems fine to me.

**Comment by atheurer** [comment]
I agree.  The problem with the first example is that is has very limited function.  

Here is an example of the highly nested YAML to replicate this command one may use today:

<pre>
pbench-uperf --test-types=stream,rr --message-sizes=256,1024 --protocols=tcp,udp --instances=1,8 
</pre>

<pre>
- test-types:
  - stream, rr:
    - message-sizes:
      - 256, 1024:
        - protocols:
          - tcp, udp:
            - instances:
              - tcp, udp
</pre>

Note that the above pbench-uperf command always builds a full matrix of those options, this simple example can be represented in YAML with any hierarchical order.  For example, this would produce the same set of benchmark iterations:

<pre>
- instances:
  - 1, 8:
    - protocols:
      - tcp, udp:  
        - message-sizes:
          - 256, 1024:
            - test-types:
              - stream, rr
</pre>

For each of these, there would be 2 * 2 * 2 * 2 = 16 benchmark iterations

One potential problem I am concerned about:  in these examples, when there is a list like "- 1, 8:", this is not really represented as a list (of numbers 1 and 8).  It is considered a single value.  The problem is that YAML will not allow a real list to be mapped to a value, like: "[1,8]:".  Maybe I am just doing this incorrectly.  For example, let's take the above YAML.  Right now, I have:

<pre>
- instances:
 - 1,8:
   - protocols:
    - tcp, udp
</pre>

I want to ensure that for both instance values 1 and 8, we create a benchmark iteration for every protocol listed, and not for just instance value 1 or just instance value 8.  I think the proper way to represent this would be:

<pre>
- instances:
  - 1
  - 8
  - protocols:
    - tcp
    - udp
</pre>

however, our parsing needs to understand that for instance counts 1 and 8, the 2 different protocols need to be used (and not just instance count 8).

**Comment by atheurer** [comment]
yaml parser:
[http://yaml-online-parser.appspot.com/](url)

If you have ideas of how benchmark options should work (with YAML), please try them with the parser first to ensure they work.

**Comment by atheurer** [comment]
another more complex example:

<pre>
- protocols:
    - tcp:
      - message-sizes:
        - 256
        - 1024
        - instances:
          - 16:
            - test-types:
              - stream
              - maerts
          - 8:
            - test-types:
              - bidirec
        - 16384
        - instances:
          - 2:
            - test-types:
              - stream
              - maerts
          - 1:
            - test-types:
              - bidirec
    - udp:
      - message-sizes:
        - 1024:
          - instances:
            - 1:
            - 1024:
              - test-types:
                - rr


</pre>

**Comment by atheurer** [comment]
@bengland2 can you take a look at this fio example and comment?

A set of benchmark iterations that uses different io-options for different test-types could use:
<pre>
- block-sizes:
  - 4KiB
  - 64KiB
  - 1024KiB
  - io-engine:
    - libaio:
      - targets:
        - /path/to/my/test-filesystem-or-device
        - test-types:
          - read:
          - randread:
            - io-options:
               - direct,sync
          - write:
            - io-options:
              - fsync_on_close
              - direct
              - direct,sync
          - randwrite:
            -io-options:
              - direct
</pre>
  
This would run the following benchmark iterations:

4KiB-libaio-test_dir-read-direct,sync
4KiB-libaio-test_dir-randread-direct,sync
64KiB-libaio-test_dir-read-direct,sync
64KiB-libaio-test_dir-randread-direct,sync
1024KiB-libaio-test_dir-read-direct,sync
1024KiB-libaio-test_dir-randread-direct,sync
4KiB-libaio-test_dir-write-fsync-on-close
4KiB-libaio-test_dir-write-direct
4KiB-libaio-test_dir-write-direct,sync
64KiB-libaio-test_dir-write-fsync-on-close
64KiB-libaio-test_dir-write-direct
64KiB-libaio-test_dir-write-direct,sync
1024KiB-libaio-test_dir-write-fsync-on-close
1024KiB-libaio-test_dir-write-direct
1024KiB-libaio-test_dir-write-direct,sync
4KiB-libaio-test_dir-randwrite-direct
64KiB-libaio-test_dir-randwrite-direct
1024KiB-libaio-test_dir-randwrite-direct

**Comment by bengland2** [comment]
@atheurer the basic idea is fine, sorry I didn't see your post when it happened.   So I would do it like this:
```
- block-sizes:
  - 4KiB
  - 64KiB
  - 1024KiB
- io-engine:
  - libaio:
- targets:
   - /path/to/my/test-filesystem-or-device
- test-types:
   - read:
   - randread:
     - io-options:
       - direct,sync
   - write:
     - io-options:
       - fsync_on_close
        - direct
        - direct,sync
   - randwrite:
      - io-options:
        - direct
```
where this was indented

**Comment by bengland2** [comment]
@atheurer, I didn't finish that sentence ;-( So I'm doing the same basic YAML lines as you but the indentation is different.  io-engine should be at same level as block-sizes, targets, and test-types.  We're implying combinations of parameters used to generate a set of test runs, by putting multi-valued parameters at the same level.

**Comment by atheurer** [comment]
@bengland2 I am trying to avoid the indentation you have because I want the most flexibility.  For example, I don't want ioengine at the same indentation because I may want to run two different ioengines for the same set of block sizes, like this:

<pre>
- block-sizes:
  - 4KiB
  - 64KiB
  - 1024KiB
  - io-engine:
    - sync
      - targets:
        - /dev/sdb
      - test-types
          - write:
            - io-options:
              - fsync_on_close
              - direct
              - direct,sync
          - randwrite:
            -io-options:
              - direct
    - libaio:
      - targets:
        - /path/to/my/test-filesystem-or-device
        - test-types:
          - read:
          - randread:
            - io-options:
               - direct,sync
</pre>

This implies that there are no rules on what has to have the same indentation.  You just have to traverse the list like a tree to find all of the iterations.  For example, instead of starting with --block-sizes as the first and left-most line, you could start with test-types:, or really anything else.  

If you want, you can have another set of nested parameters that start all the way to the left again:

<pre>
- block-sizes:
  - 4KiB
  - test-types:
    - read
- block-sizes:
  - 8KiB
  - test-types:
    - write
</pre>

So, in that description, only 4KiB is tested for read and only 8Kib for write
